### ISSUE TITLE
Request Context works from callers to callees but not in the reverse direction

### ISSUE THREAD

**Issue by garbervetsky** [dotnet/orleans#746]
In this direction it works:

Client code {
  RequestContext.Set("GrainInfo", 10L);
  grainRef.M()
}

Grain code for M {
          var clientData = (long)RequestContext.Get("GrainInfo");
}

But this fails:

Grain code for M{
  RequestContext.Set("GrainInfo", 10L);
  return;
}

Client code {
    grainRef.M();
   var clientData = (long)RequestContext.Get("GrainInfo"); // Null pointer
}

The tests demonstrating this problem are in: https://github.com/dotnet/orleans/pull/747 and https://github.com/dotnet/orleans/pull/748.

**Comment by gabikliot** [comment]
I am looking into it.
It appears to be more complicated than what I initially thought.
This feature basically requires `CallContext.LogicalGetData` to flow "up" (back) the chain call, also via `TaskCompletionSource` and I am not sure if it can work this way. 
Investigating with TPL Gods (@stephentoub)

**Comment by veikkoeeva** [comment]
@gabikliot I'm interested to hear about this too. I have a faint memory poking around this years ago. Hmm, I wonder if these are of help (on the hoof):
* [CallContext.LogicalGetData Vs. CallContext.GetData](https://stackoverflow.com/questions/6097301/callcontext-logicalgetdata-vs-callcontext-getdata) (the dicussion and pointers may be more worth than the question itself).
* [Logical Call Context: Flowing Data across Threads, AppDomains, and Processes](http://www.wintellect.com/devcenter/jeffreyr/logical-call-context-flowing-data-across-threads-appdomains-and-processes)
* [Implicit Async Context ("AsyncLocal")](http://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html)

**Comment by gabikliot** [comment]
@veikkoeeva , thanks for the links. I am aware of all of them. Number 3 is super useful since it explains about the lazy copy-on-write behavior of CallContext. But indeed, they are all talking about forward flowing, while we need backward via TaskCompletionSource. I asked Stephen Taub directly.

**Comment by gabikliot** [comment]
OK, so I checked about that issue. There appears to be 2 complications:

1) **Semantics** - if we use the `RequestContext` to flow back the responses, we need to think about immutability of `RequestContext`. What if the down call decided to delete an item from `RequestContext`, would that mean that after this call is awaited we need to remove the item from the caller's `RequestContext` as well? That does not sound right. It looks like the proper semantics should be to define another concept `ResponsetContext`. While `RequestContext` flows only downstream, `ResponsetContext` flows only upstream and they don't share anything. Totally doable, just need to agree on that new concept.

2) **Implementation** - there appears to be a major complication with Implementation of `ResponsetContext`. To understand why, let me briefly explain how we implement the `RequestContext` now . We store the actual data in what is called [`CallContext.LogicalSetData`](https://github.com/dotnet/orleans/blob/master/src/Orleans/Runtime/RequestContext.cs#L178). This CallContext.LogicalSetData is kind of like thread local storage, but logical. It means it is local on the executing thread, but will flow **automatically** between threads and Tasks when Tasks are switched (will through via `awaits`). That is exactly what we want from `RequestContext` - when grain call starts we set the `RequestContext` that was stored in the message header into `CallContext.LogicalSetData` inside a grain. Now TPL will automatically flow it across `await` boundaries and it will be available in any place within this grain method call (including some library call that may not even know that it is running inside a grain, but can just retrieve that data from either `RequestContext` or `CallContext.LogicalGetData`). So far so good.
For  `ResponsetContext` we want to do the same. When the grain call is awaited, we want to set the `ResponsetContext` from the message header into `CallContext.LogicalSetData` and make it available to the rest of the grain code. But here is the problem - we can't do that. 
When we make a grain call we create `TaskCompletionSource` and return `TaskCompletionSource.Task` back to the caller code that awaits it. When the response arrives (on a different thread of course) we resolve that `TaskCompletionSource`. At this point we also would like to "inject" the `ResponsetContext` content from the message into the resolved Task. But that is not possible. There’s no way to flow additional context along with a TaskCompletionSource.Set* call that would then be available ambiently when its Task was awaited.
**What that means is, at least with the current Orleans RPC architecture, it is not possible, as far as I can tell, to implement `ResponsetContext` via `CallContext.LogicalSetData`.**
It is possible of course to implement it differently, maybe via some other Thread Local, or for example by setting the `ResponsetContext` on some grain member field. But that would mean that `ResponsetContext` semantics (w.r.t to flowing via await boundaries and the accessibility from any where and maybe more) will differ from `RequestContext`.

Bottom line - it looks like implementing the reverse direction of `RequestContext` with the same semantics and behavior as a direct direction may not be possible. Therefore, we need to think deeper about what behavior we want to this `ResponsetContext` and how to implement it differently.

We can leave this issue open, or close for now, but unless someone steps up with good ideas, I don't see us investing much into this at this point.

**Comment by garbervetsky** [comment]
Hi

1) I agree about the new name.

2) I was thinking in something different but it may require to do some
"compile" time actions. One option would be:

- Have a "ghost" version of the grain methods that returns the same type as
the real method but with a wrapper that contains a Response object. That is
in returns Task<Wrapper<T>>

- The ghost method includes the original method the logic to set put the
response context in the returned object

- At the client side the "codegen" invokes the ghost method that recover
the original return value and the response context

One possible optimization: if we can determine statically (or dynamicalle)
the the grain that not contain a response context we can do a direct call
to the original method.

Make sense?

Thanks!

**Comment by gabikliot** [comment]
Yes, that is definitely one option. Lots of overhead and work for this one feature (code gen is complicated), and there might be easier ways to achieve this (for example, by setting the response context in the grain member field), but what you suggested is also possible.

**Comment by gabikliot** [comment]
Two extra links from [Gitter](https://gitter.im/dotnet/orleans?at=5605a140131b784f781fec1a), just for reference:
https://stackoverflow.com/questions/31707362/how-do-the-semantics-of-asynclocal-differ-from-the-logical-call-context
https://stackoverflow.com/questions/31953846/if-i-cant-use-tls-in-c-sharp-async-programming-what-can-i-use

**Comment by garbervetsky** [comment]
Hi. Closed means this is not going to be implemented. Isn't it?

**Comment by sergeybykov** [comment]
I though we decided to "won't fix" it because of the complexity while there are easier ways to achieve the same.

**Comment by garbervetsky** [comment]
OK! ;-)

Thanks
